Circle CI: Run in Ubuntu and macOS environments, in parallel

janpio opened the pull request:
This PR takes the current Circle CI configuration for the macOS test run and creates a copy for Ubuntu, makes all the necessary changes (some commands have to be replaced, a dependency that is always present on macOS has to be manually installed, the caches need better names) and then adds a "workflow" that makes both test, macOS and Ubuntu, run at the same time.

janpio committed:
Switch from macOS to Ubuntu-Ruby-Docker image

janpio committed:
new cache key as old cache can't be restored

janpio committed:
no brew on ubuntu, use apt-get

janpio committed:
change default/ENV encoding

janpio committed:
set FASTLANE_ITUNES_TRANSPORTER_PATH to existing path (of non existing iTMSTransporter installation)

janpio committed:
download, build and install xar

janpio committed:
combine macos and ubuntu config.yml
name jobs
add workflows to parallelize both

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Pull Request is Work in Progre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

janpio commented:
Success on first try:

![image](https://user-images.githubusercontent.com/183673/34065270-9e3a782e-e200-11e7-84e1-07911112f2fd.png)

👍

janpio commented:
Although Github UI still shows this:
![image](https://user-images.githubusercontent.com/183673/34065294-c9db1290-e200-11e7-8a19-3c3e674d231e.png)
Bug or did I miss something?

janpio commented:
Found something, seems to be a limitation of Github/Circle integration:
https://circleci.com/docs/2.0/workflows/#workflows-waiting-for-status-in-github
We can change that when this PR is merged.

KrauseFx commented:
Wow 🚀

janpio commented:
PR is now ready for review.

KrauseFx commented:
<img width="699" alt="screenshot 2017-12-19 08 56 43" src="https://user-images.githubusercontent.com/869950/34146491-8e31571c-e49a-11e7-86b2-bd70e3c182ab.png">

Looks like CI is green, but the overall status is missing or something?

KrauseFx commented:
This looks awesome, and I'm really excited to have Linux CI running also. Since this is a very Circle heavy PR, I'd love to get @DanToml's input on this before we merge. I know she's on vacation right now, so if it's cool from your side, let's wait for her to come back and get her input?

janpio commented:
Absolutely, looking forward to it - this was my first contact with Circle CI.

Further work that benefits from Ubuntu or cross platform testing can just use this branch as base instead of `master`, so no harm done.

janpio commented:
Oh, missed your second comment with the screenshot: Yes, that what I talked about in https://github.com/fastlane/fastlane/pull/11242#issuecomment-352146780 - you (or some other admin) would have to change a setting here in Github for this to be correct.

KrauseFx commented:
Yeah, seems like it's something we should wait for until after the merge

janpio commented:
Yes, as we know how to solve it (and CircleCI is aware and has this explicit instructions) I don't think this will become a problem to solve, we just shouldn't forget to do it when we merged this.

lacostej commented on .circleci/config.yml:
```diff
@@ -23,7 +23,61 @@ jobs:
+            cd /tmp
+            mkdir download
+            cd download
+            wget https://github.com/downloads/mackyle/xar/xar-1.6.1.tar.gz
```
We might want to provide a pre-built binary to avoid having to do that every time and maybe host it here: https://launchpad.net/ubuntu/+ppas

You could of course host it on github first.

janpio replied:
That sounds like a good idea. I will look into that.

janpio replied:
Fascinating, there are actually 4 sources for ready built `xar`s there I didn't find googling before:
https://launchpad.net/ubuntu/+ppas?name_filter=xar
Unfortunately none of them are v1.6.1, so I will see if I can build this myself.

janpio replied:
https://launchpad.net/ubuntu/+ppas links to https://help.launchpad.net/Packaging/PPA links to https://help.launchpad.net/Packaging/PPA/Uploading links to https://help.launchpad.net/Packaging/PPA/BuildingASourcePackage links to http://packaging.ubuntu.com/html/ links to http://packaging.ubuntu.com/html/packaging-new-software.html - and is a bit complicated.

I think I will move this into a new issue and work on it independently from this PR. 

Do you think deploying it like this would be fine for the beginning @lacostej ?

janpio replied:
https://github.com/fastlane/fastlane/issues/11280 now exists.

DanToml replied:
We should build a docker container that includes these (based on the circleci/ruby:2.3 container)

janpio replied:
See #11302

lacostej commented on .circleci/config.yml:
```diff
@@ -23,7 +23,61 @@ jobs:
+            make
+            sudo make install
+      - save_cache:
+          key: fastlane-ubuntu-gems-{{ checksum "Gemfile" }}
```
does /Users/distiller work on Linux? (store_test_results section)

janpio replied:
afaik it doesn't even work on macOS - or I couldn't find out how it should work, what it should do etc. Do you have more insight?

DanToml replied:
/Users/distiller will not work on linux

janpio replied:
Ok, looked into this a bit:

This step is currently failing on our normal builds:
https://circleci.com/gh/fastlane/fastlane/12636
```
Unable to save test results from /Users/distiller/test-results
Error stat /Users/distiller/test-results: no such file or directory
Found no path with test results, skipping
```

But test results should actually be written to somewhere: https://github.com/fastlane/fastlane/blob/3830c51f153875ec0b7350af359709e32616cfab/Rakefile#L18

I will create an issue for this first, when it is fixed the result can be incorporated here.

janpio replied:
This is now #11303 and PR #11304

janpio replied:
PR #11304 is ready for review (and merge). It fixes this step and will probably work unchanged on Ubuntu as well, so can be copied over.

DanToml commented on .circleci/config.yml:
```diff
@@ -1,7 +1,7 @@
 
 jobs:
-  build:
+  "macOS":
```
no need to wrap these in quotes

janpio replied:
I wanted to indicate that these are chosen strings vs. stuff that from a list Circle expects. Bad idea?

DanToml replied:
They're both the same when they get parsed anyway :woman_shrugging:, as much as I wish YAML required quotes everywhere, stylistically, I'm not a fan of them when they're not required, as it makes me assume something weird is going on.

DanToml commented on .circleci/config.yml:
```diff
@@ -13,7 +13,7 @@ jobs:
 
       - restore_cache:
-          key: v1-gems-{{ checksum "Gemfile" }}
+          key: fastlane-macos-gems-{{ checksum "Gemfile" }}
```
`v1-macos-gems-{{ checksum "Gemfile" }}` would be a better cache key here, because we need to be able to invalidate the cache key easily

janpio replied:
Found it: https://circleci.com/docs/2.0/caching/#clearing-cache

DanToml commented on .circleci/config.yml:
```diff
@@ -23,7 +23,61 @@ jobs:
+            make
+            sudo make install
+      - save_cache:
+          key: fastlane-ubuntu-gems-{{ checksum "Gemfile" }}
```
`v1-ubuntu-gems-{{ checksum "Gemfile" }}` would be a better cache key here, because we need to be able to invalidate the cache key easily

janpio replied:
`v1` refers to what here? Is this v1 of your cache file? I was unclear if this references Circle CI 1.0 maybe somehow.

DanToml replied:
Yeah - We don't have a good way to clear caches through the UI/API yet, so the hack is to bump a version string in the yaml

DanToml commented on .circleci/config.yml:
```diff
@@ -23,7 +23,61 @@ jobs:
+          name: Setup Build
+          command: |
+            mkdir -p $CIRCLE_TEST_RESULTS
+            echo "2.3" > .ruby-version
```
No need to write .ruby-version on linux

janpio replied:
Got it.

DanToml commented on .circleci/config.yml:
```diff
@@ -23,7 +23,61 @@ jobs:
+          command: |
+            mkdir -p $CIRCLE_TEST_RESULTS
+            echo "2.3" > .ruby-version
+            gem update --system
```
No need to update system rubygems on linux either, the circle image builds nightly

janpio replied:
Got it.

DanToml commented on .circleci/config.yml:
```diff
@@ -23,7 +23,61 @@ jobs:
+            echo "2.3" > .ruby-version
+            gem update --system
+            bundle check || bundle install --jobs=4 --retry=3 --path .bundle
+            sudo apt-get install shellcheck
```
Another candidate for the build docker container.
Ideally we should seperate shellcheck linting out and run it in its own job.

janpio replied:
See #11302

DanToml replied:
:+1:

DanToml commented on .circleci/config.yml:
```diff
@@ -23,7 +23,61 @@ jobs:
+      FASTLANE_ITUNES_TRANSPORTER_PATH: .bundle
+    docker:
+      - image: circleci/ruby:2.3
+    shell: /bin/bash --login -eo pipefail
```
Probably don't need a login shell on linux?

janpio replied:
Will look into what this actually does/means. Stupid copy/paste from macOS for now.

janpio replied:
(Why do we need this on macOS?)

janpio replied:
Read up on it:
Don't think we need that.

https://circleci.com/docs/2.0/configuration-reference/#default-shell-options


(Why do we need this on macOS?)

janpio replied:
Read up on it:
Don't think we need that.

https://circleci.com/docs/2.0/configuration-reference/#default-shell-options

(Why do we need this on macOS?)

janpio replied:
Read up on it: https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html#Bash-Startup-Files
Don't think we need that.

(Why do we need this on macOS?)

When `--login` is gone, all that is left is actually the default: https://circleci.com/docs/2.0/configuration-reference/#default-shell-options
So this line can be completely removed 👍

DanToml replied:
We need it on macOS jobs so that we source the `chruby` init scripts, to use our chosen ruby, and not system ruby

endocrimes reviewed (changes requested):
This is a good start!

KrauseFx commented:
Thanks for your help here @DanToml ❤️

janpio committed:
better cache keys
moved installation of additional dependency outside of cache "bracket" for more logic

janpio committed:
login shell not necessary on ubuntu

janpio committed:
ruby version not necessary on ubuntu

janpio committed:
updating gems not necessary on ubuntu

janpio commented on .circleci/config.yml:
```diff
@@ -23,10 +23,62 @@ jobs:
+      - run: bundle exec fastlane test
+
+      - store_test_results:
+          path: /Users/distiller/test-results
```
replace with result from #11302

janpio commented:
Update:
- Waiting for review and merge of #11302 
- Dependency stuff that should be solved via a Docker container was "outsourced" to #11302

Besides that I hope I covered all comments.

janpio commented:
#11302 is merged, I will incorporate this here now.

janpio committed:
fix test reports stuff again (as only half the file got it at rebase)

janpio committed:
rename "test-results" to "test-reports" (which corresponds to the suggested Circle CI env var) to avoid future confusion

janpio commented:
Rebased, and changes from #11304 adapted for the second half of the file. Circle CI now correctly collects the test reports of rspec and display them:

> Your build ran 4541 tests in RSpec with 0 failures
Slowest test: fastlane_core.spec.build_watcher_spec FastlaneCore::BuildWatcher.wait_for_build_processing_to_be_complete returns specified build when strict watch is enabled (took 10.00 seconds).

I think this is now "good" again and all relevant, outstanding feedback was addressed. Can I get another look at it please?

(As said before, I think the dependency and Docker stuff will be handled in a follow up issue: #11302)

DanToml commented on .circleci/config.yml:
```diff
@@ -1,45 +1,107 @@
+  build:
+    jobs:
+      - "macOS"
+      - "Ubuntu"
```
Add a newline at the end and this is good to go

DanToml commented:
@KrauseFx On merge, this will also require changing the required status' for merging, as the old `ci/circleci` named status doesn't get reported in workflows (because we get a status per job)

janpio committed:
Add newline at end of file

janpio commented:
Oh, I actually **can't** merge this before @KrauseFx changes that setting as this is one of the expected checks ;) 

See here for instructions: https://circleci.com/docs/2.0/workflows/#workflows-waiting-for-status-in-github
